src/core/compile/check/mod.rs

```diff
@@ -791,14 +791,7 @@ impl<'a> Context<'a> {
                                 vec![],
                                 Ty::Transformed(
                                     Ty::prelude(Prelude::Pubkey, vec![]).into(),
-                                    Transformation::new(|mut expr| {
-                                        let account = match1!(expr.obj, ExpressionObj::Call { function, .. } => *function);
-                                        let account = match1!(account.obj, ExpressionObj::Attribute { value, .. } => *value);
-
-                                        expr.obj = ExpressionObj::Rendered(quote! {
-                                            #account.borrow().__account__.key()
-                                        });
-
+                                    Transformation::new(|expr| { 
```

Comment:
I think that the special case you added to the build stage can be implemented here - you can use Transformation::new_with_context to create a transformation that can use the context stack.

Otherwise, if I'm wrong, you can remove this transformation entirely since it wouldn't do anything (just make `Pubkey` the return type of the `key` function instead of `Ty::Transformed(...)`).

Reply:
Thanks, that makes sense! I've tried moving the same logic into a `new_with_context` block in check:

```rs
Transformation::new_with_context(|mut expr, context_stack| {
    if !context_stack.has(&ExprContext::Seed) {
        // If not in a seed, rewrite account.key() to account.borrow().__account__.key()
        let value = ExpressionObj::BorrowImmut(expr.obj.into());
        let value = ExpressionObj::Attribute {
            value: value.into(),
            name: String::from("__account__"),
        };
        let value = ExpressionObj::Attribute {
            value: value.into(),
            name: "key".to_string(),
        };
        expr.obj = value;
    }
    Ok(Transformed::Expression(expr))
}, None)
```

But for some reason it doesn't quite work the same, and the output I'm getting (for non seed cases) looks like:

```rs
calculator.key().borrow().__account__.key
```

So when I do `expr.obj()` it's getting `calculator.key()` where in the build stage it was getting `calculator` as the base I built on

And when I do the `ExpressionObj::Attribute` at the end, it's writing out `.key` instead of `.key()` as it did in the current code

Do you know what would cause those differences/how I could fix them in the check stage?

Reply:
Okay cool I think I've got it working, and now I understand what this stuff is doing!

```rs
let account = match1!(expr.obj, ExpressionObj::Call { function, .. } => *function);
let account = match1!(account.obj, ExpressionObj::Attribute { value, .. } => *value);
```

Added those back to get to the `account` from `account.key()` and it's looking good!